src/tracing/metric_event.h

```diff
@@ -0,0 +1,181 @@
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+#pragma once
+
+#include <lib/core/CHIPError.h>
+#include <tracing/metric_keys.h>
+#include <tracing/metric_macros.h>
+#include <tracing/registry.h>
+
+namespace chip {
+namespace Tracing {
+
+/**
+ * Define a metric that can be logged. A metric consists of a key-value pair. The value is
+ * currently limited to simple scalar values.
+ *
+ * Additionally a metric is tagged as either an instant event or marked with a begin and end
+ * for the event. When the latter is used, a duration can be associated between the two events.
+ */
+class MetricEvent
+{
+public:
+    MetricEvent(const MetricEvent &)             = default;
+    MetricEvent(MetricEvent &&)                  = default;
+    MetricEvent & operator=(const MetricEvent &) = default;
+    MetricEvent & operator=(MetricEvent &&)      = default;
+
+public:
+    // This specifies the different categories of metric events that can created. In addition to
+    // emitting an event, events paired with a kBeginEvent and kEndEvent can be used to track
+    // duration for the event. A kInstantEvent represents a one shot event.
+    enum class Type
+    {
+        kBeginEvent,  // Marks the begin of an event. This must be tied with an End event.
+        kEndEvent,    // Marks the end of an event. This is typically preceeded with a Begin event.
+        kInstantEvent // No duration
+    };
+
+    // This defines the different types of values that can stored when a metric is emitted
+    struct Value
+    {
+        Value(const Value &)             = default;
+        Value(Value &&)                  = default;
+        Value & operator=(const Value &) = default;
+        Value & operator=(Value &&)      = default;
+
+        enum class Type : uint8_t
+        {
+            kInt32,        // int32_t
+            kUInt32,       // uint32_t
+            kChipErrorCode // chip::ChipError
+        };
+
+        union Store
+        {
+            int32_t int32_value;
+            uint32_t uint32_value;
+
+            Store(int32_t v) : int32_value(v) {}
+
+            Store(uint32_t v) : uint32_value(v) {}
+        };
+
+        Store store;
+        Type type;
+
+        Value(uint32_t value) : store(value), type(Type::kUInt32) {}
+
+        Value(int32_t value) : store(value), type(Type::kInt32) {}
+
+        Value(const ChipError & err) : store(err.AsInteger()), type(Type::kChipErrorCode) {}
+    };
+
+    MetricEvent(Type type, MetricKey key, int32_t value = 0) : mType(type), mKey(key), mValue(value) {}
```

Comment:
Such a default seems wrong: why is a empty value acceptable for some metric key? This seems to encourage "send a metric, however the metric content does not matter". I believe this is due to the start/end design. It seems better to split it.

Reply:
I can change to be explicit